### PR TITLE
feature/go-pfdetect

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Replace the perl version of pfdetect with go

Delete branch after merge
=========================
NO

NEWS file entries
=================

Enhancements
------------
* Replace perl pfdetect with a go version

UPGRADE file entries
=================
Any regex detect parser must be rewritten using the re2 regex which is mostly backwards compatible perl style regex

### CODE REVIEW

**Comment by jrouzierinverse** [comment]
Changes for logging done.

File: addons/packages/packetfence.spec [hunk 1]
```diff
@@ -1268,6 +1268,7 @@ fi
 %dir                    /usr/local/pf/sbin
 %attr(0755, pf, pf)     /usr/local/pf/sbin/pfbandwidthd
 %attr(0755, pf, pf)     /usr/local/pf/sbin/pfdetect
+%attr(0755, pf, pf)     /usr/local/pf/sbin/pfdetect-perl
```
Comment (julsemaan) [inline addons/packages/packetfence.spec]: I would just remove it, there is no point in keeping that old piece of code

File: go/detect/main.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+func (array *Runners) Append(r ...*ParseRunner) {
+	*array = append(*array, r...)
+}
+
+func (s *Server) AddRunner(runner *ParseRunner) {
+	s.Runners.Append(runner)
+}
+
+func (s *Server) StopRunners() {
+	log.Logger().Info("Stopping runners")
```
Comment (julsemaan) [inline go/detect/main.go]: I wouldn't really call those runners in the logging, parsers would be a better word

File: go/detect/main.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+					s.Done("Signal")
+					break
+				}
+				s.ReloadConfig()
+			}
+			s.NotifySystemd(daemon.SdNotifyStopping)
+			s.StopRunners()
+		}()
+		wg.Wait()
+		log.Logger().Info("Done signal setup")
```
Comment (julsemaan) [inline go/detect/main.go]: debug ?

File: go/detectparser/generic.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	LastIfMatch      bool
+	IpMacTranslation bool
+	Actions          []GenericParserAction
+}
+
+var macGarbageRegex = regexp.MustCompile(`[\s\-\.:]`)
+var validSimpleMacHexRegex = regexp.MustCompile(`^[a-fA-F0-9]{12}$`)
+var macPairHexRegex = regexp.MustCompile(`[a-fA-F0-9]{2}`)
+
+func cleanMac(mac string) string {
```
Comment (julsemaan) [inline go/detectparser/generic.go]: This would be better in the sharedutils package

File: go/pfconfigdriver/fetch.go [hunk 5]
```diff
@@ -257,6 +257,17 @@ func FetchDecodeSocketCache(ctx context.Context, o PfconfigObject) (bool, error)
 	return true, err
 }
 
+// Fetch the keys of a resource
```
Comment (julsemaan) [inline go/pfconfigdriver/fetch.go]: of a namespace

File: lib/pfconfig/namespaces/config/Pfdetect.pm [hunk 6]
```diff
@@ -48,12 +48,6 @@ sub build_child {
             for my $rule_id (@rule_ids) {
                 $rule_id =~ /^$id rule (.*)/;
                 my $rule = {%{$tmp_cfg{$rule_id}}, name => $1};
-                my $regex = eval {qr/$rule->{regex}/};
```
Comment (julsemaan) [inline lib/pfconfig/namespaces/config/Pfdetect.pm:51]: So that means there isn't any regex validation anymore in the admin ?

**Comment by julsemaan** [comment]
Old code of pfdetect should also be removed prior to the merge of this PR

File: lib/pfconfig/namespaces/config/Pfdetect.pm (hunk 6 above)
Comment (jrouzierinverse) [reply lib/pfconfig/namespaces/config/Pfdetect.pm:51]: There are still regex validation.
Just in the admin.

File: go/detectparser/dhcp.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+var dhcpRegexPattern2 = regexp.MustCompile(`(DHCPDISCOVER|DHCPOFFER|DHCPREQUEST|DHCPACK|DHCPRELEASE|DHCPINFORM|DHCPEXPIRE) to ([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+) \(([0-9a-f]{2}:[0-9a-f]{2}:[0-9a-f]{2}:[0-9a-f]{2}:[0-9a-f]{2}:[0-9a-f]{2})\)`)
+
+type DhcpParser struct {
+	Pattern1, Pattern2 *regexp.Regexp
+}
+
+func (s *DhcpParser) Parse(line string) ([]ApiCall, error) {
+	for _, r := range []*regexp.Regexp{s.Pattern1, s.Pattern2} {
+		if matches := r.FindStringSubmatch(line); matches != nil && matches[1] == "DHCPACK" {
+			return []ApiCall{
```
Comment (fdurand) [inline go/detectparser/dhcp.go]: Maybe just a check if the ip is really an ip (ParseIP( matches[2]))
Same for the mac address (ParseMAC(matches[3]))

File: go/detectparser/fortianalyser.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	matches := s.Pattern1.Split(line, -1)
+	var srcip, logid string
+	for _, str := range matches {
+		args := s.Pattern2.Split(str, 2)
+		if len(args) <= 1 {
+			continue
+		}
+
+		if args[0] == "srcip" {
+			srcip = args[1]
```
Comment (fdurand) [inline go/detectparser/fortianalyser.go:26]: ip validation

File: go/detectparser/generic.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+func (rule *GenericParserRule) GetReplacementMap(ctx context.Context, src string, match []int) map[string]string {
+	replacementStrings := make(map[string]string)
+	subexpNames := rule.Match.SubexpNames()
+	for i, name := range subexpNames {
+		if name != "" {
+			replacementStrings[name] = src[match[2*i]:match[2*i+1]]
+		}
+	}
+
+	mac, macFound := replacementStrings["mac"]
```
Comment (fdurand) [inline go/detectparser/generic.go:180]: validate mac

File: go/detectparser/generic.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+	ip, ipFound := replacementStrings["ip"]
```
Comment (fdurand) [inline go/detectparser/generic.go:181]: validate ip

File: go/detectparser/nexpose.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+func (s *NexposeParser) Parse(line string) ([]ApiCall, error) {
+	if matches := s.Pattern1.FindStringSubmatch(line); matches != nil && matches[4] == "VULNERABILITY" {
+		return []ApiCall{
+			&PfqueueApiCall{
+				Method: "event_add",
+				Params: []interface{}{
+					"date", matches[1],
+					"dstip", matches[3],
```
Comment (fdurand) [inline go/detectparser/nexpose.go]: validate ip

File: go/detectparser/nexpose.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+					"srcip", matches[2],
```
Comment (fdurand) [inline go/detectparser/nexpose.go]: validate ip

File: go/detectparser/security_onion.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+	if len(matches2) != 10 {
+		return nil, fmt.Errorf("Error parsing")
+	}
+
+	return []ApiCall{
+		&PfqueueApiCall{
+			Method: "event_add",
+			Params: []interface{}{
+				"date", matches1[1],
+				"srcip", matches2[0],
```
Comment (fdurand) [inline go/detectparser/security_onion.go:33]: validate ip

File: go/detectparser/security_onion.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+				"dstip", matches2[1],
```
Comment (fdurand) [inline go/detectparser/security_onion.go:34]: validate ip

File: go/detectparser/snort.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func (s *SnortParser) Parse(line string) ([]ApiCall, error) {
+	if matches := s.Pattern1.FindStringSubmatch(line); matches != nil {
+		return []ApiCall{
+			&PfqueueApiCall{
+				Method: "event_add",
+				Params: []interface{}{
+					"date", matches[1],
+					"dstip", matches[6],
```
Comment (fdurand) [inline go/detectparser/snort.go]: validate ip

File: go/detectparser/snort.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+					"srcip", matches[4],
```
Comment (fdurand) [inline go/detectparser/snort.go]: validate ip

File: go/detectparser/snort.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+		}, nil
+	}
+
+	if matches := s.Pattern2.FindStringSubmatch(line); matches != nil {
+		return []ApiCall{
+			&PfqueueApiCall{
+				Method: "event_add",
+				Params: []interface{}{
+					"date", matches[1],
+					"srcip", matches[3],
```
Comment (fdurand) [inline go/detectparser/snort.go]: validate ip

File: go/detectparser/snort.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+		}, nil
+	}
+
+	if matches := s.Pattern3.FindStringSubmatch(line); matches != nil {
+		return []ApiCall{
+			&PfqueueApiCall{
+				Method: "event_add",
+				Params: []interface{}{
+					"date", matches[1],
+					"srcip", matches[3],
```
Comment (fdurand) [inline go/detectparser/snort.go]: validate ip

File: go/detectparser/suricata_md5.go [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+	if tmp, err = s.ResolverIp2Mac.IpToMac(ip); err != nil {
+		return nil, err
+	}
+
+	if mac, ok = tmp.(string); !ok {
+		return nil, fmt.Errorf("endpoint not found")
+	}
+
+	data["mac"] = mac
```
Comment (fdurand) [inline go/detectparser/suricata_md5.go:71]: validate mac

**Review by fdurand — CHANGES_REQUESTED** [review]

File: go/detectparser/dhcp.go (hunk 7 above)
Comment (jrouzierinverse) [reply go/detectparser/dhcp.go]: Done by the regex.

File: go/detectparser/fortianalyser.go (hunk 8 above)
Comment (jrouzierinverse) [reply go/detectparser/fortianalyser.go:26]: Ok

File: go/detectparser/dhcp.go (hunk 7 above)
Comment (fdurand) [reply go/detectparser/dhcp.go]: https://i.redditmedia.com/6vYqs23NJ2z8IHqNztxdycCDG9-rtcuGPLpf6nODGxQ.jpg?s=b41acf6471f606d765646ab0c9005094

**Comment by julsemaan** [comment]
If @fdurand is happy with the changes done for his comments, then I'm good for merging

**Comment by extrafu** [comment]
@julsemaan Ready for merging?

**Comment by julsemaan** [comment]
Same status as my last comment:

>  If @fdurand is happy with the changes done for his comments, then I'm good for merging

He still has a "Changes requested" on the PR, just want to make sure I don't override that for something that is still pending to be fixed

**Review by fdurand — APPROVED** [review]

**Pull request merged**